File: myoelectric/api/typescript/src/classifier/classifier.ts
/**
 * WIA Myoelectric Gesture Classifier
 * @module classifier/classifier
 */

import { Observable, Subject, BehaviorSubject } from 'rxjs';
import { map, filter, bufferCount, scan } from 'rxjs/operators';
import {
  Gesture,
  FeatureVector,
  ClassificationResult,
  TrainingDataset,
  TrainingResult,
  LabeledSample,
  ClassifierConfig,
  ClassificationAlgorithm,
  ModelInfo,
  AlternativeClassification,
} from './types';

/**
 * Abstract base class for gesture classifiers
 */
export abstract class GestureClassifier {
  protected config: ClassifierConfig;
  protected modelInfo: ModelInfo | null = null;
  protected isLoaded: boolean = false;

  constructor(config: Partial<ClassifierConfig> = {}) {
    this.config = {
      algorithm: config.algorithm ?? ClassificationAlgorithm.LDA,
      confidenceThreshold: config.confidenceThreshold ?? 0.7,
      majorityVotingFrames: config.majorityVotingFrames ?? 3,
      minGestureDurationMs: config.minGestureDurationMs ?? 100,
    };
  }

  /**
   * Load a trained model from file
   */
  abstract loadModel(path: string): Promise<void>;

  /**
   * Save the current model to file
   */
  abstract saveModel(path: string): Promise<void>;

  /**
   * Classify a single feature vector
   */
  abstract classify(features: FeatureVector): ClassificationResult;

  /**
   * Train the classifier on a dataset
   */
  abstract train(dataset: TrainingDataset): Promise<TrainingResult>;

  /**
   * Fine-tune the model with additional samples
   */
  abstract finetune(samples: LabeledSample[]): Promise<void>;

  /**
   * Get model information
   */
  getModelInfo(): ModelInfo | null {
    return this.modelInfo;
  }

  /**
   * Check if model is loaded
   */
  isModelLoaded(): boolean {
    return this.isLoaded;
  }

  /**
   * Classify a stream of feature vectors with majority voting
   */
  classifyStream(
    stream: Observable<FeatureVector>
  ): Observable<ClassificationResult> {
    return stream.pipe(
      map((features) => this.classify(features)),
      bufferCount(this.config.majorityVotingFrames, 1),
      map((results) => this.majorityVote(results)),
      filter((result) => result.confidence >= this.config.confidenceThreshold)
    );
  }

  /**
   * Apply majority voting to a buffer of results
   */
  protected majorityVote(results: ClassificationResult[]): ClassificationResult {
    const voteCounts = new Map<Gesture, number>();
    const confidenceSums = new Map<Gesture, number>();

    for (const result of results) {
      const currentCount = voteCounts.get(result.gesture) ?? 0;
      const currentConf = confidenceSums.get(result.gesture) ?? 0;
      voteCounts.set(result.gesture, currentCount + 1);
      confidenceSums.set(result.gesture, currentConf + result.confidence);
    }

    let maxVotes = 0;
    let winningGesture = Gesture.REST;

    for (const [gesture, votes] of voteCounts) {
      if (votes > maxVotes) {
        maxVotes = votes;
        winningGesture = gesture;
      }
    }

    const avgConfidence =
      (confidenceSums.get(winningGesture) ?? 0) / maxVotes;
    const avgLatency =
      results.reduce((sum, r) => sum + r.latencyMs, 0) / results.length;

    // Build alternatives from other gestures
    const alternatives: AlternativeClassification[] = [];
    for (const [gesture, votes] of voteCounts) {
      if (gesture !== winningGesture) {
        alternatives.push({
          gesture,
          confidence: (confidenceSums.get(gesture) ?? 0) / votes,
        });
      }
    }

    alternatives.sort((a, b) => b.confidence - a.confidence);

    return {
      gesture: winningGesture,
      confidence: avgConfidence,
      alternatives,
      latencyMs: avgLatency,
      timestamp: results[results.length - 1].timestamp,
    };
  }
}

/**
 * Linear Discriminant Analysis classifier
 */
export class LDAClassifier extends GestureClassifier {
  private weights: number[][] = [];
  private means: number[][] = [];
  private classes: Gesture[] = [];
  private grandMean: number[] = [];

  constructor(config: Partial<ClassifierConfig> = {}) {
    super({ ...config, algorithm: ClassificationAlgorithm.LDA });
  }

  async loadModel(path: string): Promise<void> {
    // In a real implementation, load from file
    // For now, just mark as loaded
    console.log(`Loading LDA model from ${path}`);
    this.isLoaded = true;
  }

  async saveModel(path: string): Promise<void> {
    if (!this.isLoaded) {
      throw new Error('No model to save');
    }
    console.log(`Saving LDA model to ${path}`);
  }

  classify(features: FeatureVector): ClassificationResult {
    const startTime = performance.now();

    if (!this.isLoaded || this.weights.length === 0) {
      return this.defaultResult(startTime);
    }

    const x = features.features;
    const scores: number[] = [];

    // Compute discriminant scores for each class
    for (let i = 0; i < this.classes.length; i++) {
      let score = 0;
      for (let j = 0; j < x.length; j++) {
        score += this.weights[i][j] * (x[j] - this.grandMean[j]);
      }
      scores.push(score);
    }

    // Softmax to get probabilities
    const maxScore = Math.max(...scores);
    const expScores = scores.map((s) => Math.exp(s - maxScore));
    const sumExp = expScores.reduce((a, b) => a + b, 0);
    const probabilities = expScores.map((e) => e / sumExp);

    // Find best class
    let maxProb = 0;
    let bestIdx = 0;
    for (let i = 0; i < probabilities.length; i++) {
      if (probabilities[i] > maxProb) {
        maxProb = probabilities[i];
        bestIdx = i;
      }
    }

    // Build alternatives
    const alternatives: AlternativeClassification[] = [];
    for (let i = 0; i < this.classes.length; i++) {
      if (i !== bestIdx) {
        alternatives.push({
          gesture: this.classes[i],
          confidence: probabilities[i],
        });
      }
    }
    alternatives.sort((a, b) => b.confidence - a.confidence);

    return {
      gesture: this.classes[bestIdx],
      confidence: maxProb,
      alternatives,
      latencyMs: performance.now() - startTime,
      timestamp: features.timestamp,
    };
  }

  async train(dataset: TrainingDataset): Promise<TrainingResult> {
    const startTime = Date.now();

    // Group samples by class
    const classSamples = new Map<Gesture, number[][]>();
    for (const sample of dataset.samples) {
      const existing = classSamples.get(sample.label) ?? [];
      existing.push(sample.features.features);
      classSamples.set(sample.label, existing);
    }

    this.classes = Array.from(classSamples.keys());
    const numFeatures = dataset.samples[0].features.features.length;

    // Compute class means
    this.means = [];
    for (const gesture of this.classes) {
      const samples = classSamples.get(gesture)!;
      const mean = new Array(numFeatures).fill(0);
      for (const sample of samples) {
        for (let i = 0; i < numFeatures; i++) {
          mean[i] += sample[i];
        }
      }
      for (let i = 0; i < numFeatures; i++) {
        mean[i] /= samples.length;
      }
      this.means.push(mean);
    }

    // Compute grand mean
    this.grandMean = new Array(numFeatures).fill(0);
    for (const mean of this.means) {
      for (let i = 0; i < numFeatures; i++) {
        this.grandMean[i] += mean[i];
      }
    }
    for (let i = 0; i < numFeatures; i++) {
      this.grandMean[i] /= this.classes.length;
    }

    // Simplified LDA: use class means as weights
    this.weights = this.means.map((mean) =>
      mean.map((v, i) => v - this.grandMean[i])
    );

    this.isLoaded = true;

    // Compute training accuracy
    let correct = 0;
    for (const sample of dataset.samples) {
      const result = this.classify(sample.features);
      if (result.gesture === sample.label) {
        correct++;
      }
    }

    const accuracy = correct / dataset.samples.length;

    return {
      success: true,
      accuracy,
      gestureMetrics: new Map(),
      confusionMatrix: [],
      trainingDurationSec: (Date.now() - startTime) / 1000,
      modelSizeBytes: JSON.stringify(this.weights).length,
    };
  }

  async finetune(samples: LabeledSample[]): Promise<void> {
    // Simple fine-tuning: update means with new samples
    for (const sample of samples) {
      const classIdx = this.classes.indexOf(sample.label);
      if (classIdx >= 0) {
        const x = sample.features.features;
        for (let i = 0; i < x.length; i++) {
          this.means[classIdx][i] =
            0.9 * this.means[classIdx][i] + 0.1 * x[i];
        }
      }
    }

    // Update weights
    this.weights = this.means.map((mean) =>
      mean.map((v, i) => v - this.grandMean[i])
    );
  }

  private defaultResult(startTime: number): ClassificationResult {
    return {
      gesture: Gesture.REST,
      confidence: 0,
      alternatives: [],
      latencyMs: performance.now() - startTime,
      timestamp: Date.now(),
    };
  }
}

/**
 * K-Nearest Neighbors classifier
 */
export class KNNClassifier extends GestureClassifier {
  private k: number = 5;
  private trainingSamples: LabeledSample[] = [];

  constructor(config: Partial<ClassifierConfig> = {}, k: number = 5) {
    super({ ...config, algorithm: ClassificationAlgorithm.KNN });
    this.k = k;
  }

  async loadModel(path: string): Promise<void> {
    console.log(`Loading KNN model from ${path}`);
    this.isLoaded = true;
  }

  async saveModel(path: string): Promise<void> {
    console.log(`Saving KNN model to ${path}`);
  }

  classify(features: FeatureVector): ClassificationResult {
    const startTime = performance.now();

    if (this.trainingSamples.length === 0) {
      return {
        gesture: Gesture.REST,
        confidence: 0,
        alternatives: [],
        latencyMs: performance.now() - startTime,
        timestamp: features.timestamp,
      };
    }

    // Compute distances to all training samples
    const distances: { distance: number; label: Gesture }[] = [];
    for (const sample of this.trainingSamples) {
      const dist = this.euclideanDistance(
        features.features,
        sample.features.features
      );
      distances.push({ distance: dist, label: sample.label });
    }

    // Sort by distance and take k nearest
    distances.sort((a, b) => a.distance - b.distance);
    const nearest = distances.slice(0, this.k);

    // Vote among k nearest
    const votes = new Map<Gesture, number>();
    for (const neighbor of nearest) {
      const current = votes.get(neighbor.label) ?? 0;
      votes.set(neighbor.label, current + 1);
    }

    // Find winner
    let maxVotes = 0;
    let winner = Gesture.REST;
    for (const [gesture, count] of votes) {
      if (count > maxVotes) {
        maxVotes = count;
        winner = gesture;
      }
    }

    const confidence = maxVotes / this.k;

    // Build alternatives
    const alternatives: AlternativeClassification[] = [];
    for (const [gesture, count] of votes) {
      if (gesture !== winner) {
        alternatives.push({ gesture, confidence: count / this.k });
      }
    }
    alternatives.sort((a, b) => b.confidence - a.confidence);

    return {
      gesture: winner,
      confidence,
      alternatives,
      latencyMs: performance.now() - startTime,
      timestamp: features.timestamp,
    };
  }

  async train(dataset: TrainingDataset): Promise<TrainingResult> {
    const startTime = Date.now();
    this.trainingSamples = [...dataset.samples];
    this.isLoaded = true;

    // Compute accuracy using leave-one-out
    let correct = 0;
    for (let i = 0; i < dataset.samples.length; i++) {
      const test = dataset.samples[i];
      const tempSamples = this.trainingSamples;
      this.trainingSamples = [
        ...dataset.samples.slice(0, i),
        ...dataset.samples.slice(i + 1),
      ];
      const result = this.classify(test.features);
      if (result.gesture === test.label) {
        correct++;
      }
      this.trainingSamples = tempSamples;
    }

    return {
      success: true,
      accuracy: correct / dataset.samples.length,
      gestureMetrics: new Map(),
      confusionMatrix: [],
      trainingDurationSec: (Date.now() - startTime) / 1000,
      modelSizeBytes: JSON.stringify(this.trainingSamples).length,
    };
  }

  async finetune(samples: LabeledSample[]): Promise<void> {
    this.trainingSamples.push(...samples);
  }

  private euclideanDistance(a: number[], b: number[]): number {
    let sum = 0;
    for (let i = 0; i < a.length; i++) {
      const diff = a[i] - b[i];
      sum += diff * diff;
    }
    return Math.sqrt(sum);
  }
}

/**
 * Factory function to create a classifier
 */
export function createClassifier(
  algorithm: ClassificationAlgorithm,
  config: Partial<ClassifierConfig> = {}
): GestureClassifier {
  switch (algorithm) {
    case ClassificationAlgorithm.LDA:
      return new LDAClassifier(config);
    case ClassificationAlgorithm.KNN:
      return new KNNClassifier(config);
    default:
      throw new Error(`Unsupported algorithm: ${algorithm}`);
  }
}

File: myoelectric/api/rust/src/classifier.rs
//! Gesture Classification Module
//!
//! This module provides gesture classification for EMG-based
//! myoelectric prosthetic control.

use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::time::Instant;

/// Supported hand gestures
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum Gesture {
    Rest,
    HandOpen,
    HandClose,
    WristFlexion,
    WristExtension,
    WristPronation,
    WristSupination,
    PinchThumbIndex,
    PinchThumbMiddle,
    PointIndex,
    ThumbsUp,
}

impl Gesture {
    /// Get all basic gestures for simple prosthetic control
    pub fn basic_gestures() -> Vec<Gesture> {
        vec![
            Gesture::Rest,
            Gesture::HandOpen,
            Gesture::HandClose,
            Gesture::WristFlexion,
            Gesture::WristExtension,
        ]
    }

    /// Get all supported gestures
    pub fn all_gestures() -> Vec<Gesture> {
        vec![
            Gesture::Rest,
            Gesture::HandOpen,
            Gesture::HandClose,
            Gesture::WristFlexion,
            Gesture::WristExtension,
            Gesture::WristPronation,
            Gesture::WristSupination,
            Gesture::PinchThumbIndex,
            Gesture::PinchThumbMiddle,
            Gesture::PointIndex,
            Gesture::ThumbsUp,
        ]
    }

    /// Convert to string representation
    pub fn as_str(&self) -> &'static str {
        match self {
            Gesture::Rest => "rest",
            Gesture::HandOpen => "hand_open",
            Gesture::HandClose => "hand_close",
            Gesture::WristFlexion => "wrist_flexion",
            Gesture::WristExtension => "wrist_extension",
            Gesture::WristPronation => "wrist_pronation",
            Gesture::WristSupination => "wrist_supination",
            Gesture::PinchThumbIndex => "pinch_thumb_index",
            Gesture::PinchThumbMiddle => "pinch_thumb_middle",
            Gesture::PointIndex => "point_index",
            Gesture::ThumbsUp => "thumbs_up",
        }
    }
}

/// Feature vector for classification
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FeatureVector {
    /// Timestamp in milliseconds
    pub timestamp: u64,
    /// Number of channels
    pub channel_count: usize,
    /// Feature names
    pub feature_names: Vec<String>,
    /// Feature values
    pub features: Vec<f32>,
}

impl FeatureVector {
    /// Create a new feature vector
    pub fn new(timestamp: u64, features: Vec<f32>, channel_count: usize) -> Self {
        let feature_names = Self::generate_feature_names(channel_count);
        Self {
            timestamp,
            channel_count,
            feature_names,
            features,
        }
    }

    fn generate_feature_names(channels: usize) -> Vec<String> {
        let mut names = Vec::new();
        for ch in 0..channels {
            names.push(format!("ch{}_mav", ch));
            names.push(format!("ch{}_rms", ch));
            names.push(format!("ch{}_wl", ch));
            names.push(format!("ch{}_zc", ch));
            names.push(format!("ch{}_ssc", ch));
        }
        names
    }
}

/// Classification result
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ClassificationResult {
    /// Primary gesture
    pub gesture: Gesture,
    /// Confidence score (0.0 - 1.0)
    pub confidence: f32,
    /// Alternative classifications
    pub alternatives: Vec<(Gesture, f32)>,
    /// Classification latency in milliseconds
    pub latency_ms: f32,
    /// Timestamp
    pub timestamp: u64,
}

/// Classifier configuration
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ClassifierConfig {
    /// Confidence threshold for valid classification
    pub confidence_threshold: f32,
    /// Number of frames for majority voting
    pub majority_voting_frames: usize,
    /// Minimum gesture duration in milliseconds
    pub min_gesture_duration_ms: u64,
}

impl Default for ClassifierConfig {
    fn default() -> Self {
        Self {
            confidence_threshold: 0.7,
            majority_voting_frames: 3,
            min_gesture_duration_ms: 100,
        }
    }
}

/// Linear Discriminant Analysis classifier
pub struct LDAClassifier {
    _config: ClassifierConfig,
    weights: Vec<Vec<f32>>,
    grand_mean: Vec<f32>,
    classes: Vec<Gesture>,
    is_trained: bool,
}

impl LDAClassifier {
    /// Create a new LDA classifier
    pub fn new(config: ClassifierConfig) -> Self {
        Self {
            _config: config,
            weights: Vec::new(),
            grand_mean: Vec::new(),
            classes: Vec::new(),
            is_trained: false,
        }
    }

    /// Train the classifier
    pub fn train(
        &mut self,
        samples: &[Vec<f32>],
        labels: &[usize],
        gestures: Vec<Gesture>,
    ) -> Result<TrainingResult, ClassifierError> {
        if samples.is_empty() {
            return Err(ClassifierError::EmptyTrainingSet);
        }

        let start = Instant::now();
        self.classes = gestures;
        let num_features = samples[0].len();
        let num_classes = self.classes.len();

        // Group samples by class
        let mut class_samples: Vec<Vec<&Vec<f32>>> = vec![Vec::new(); num_classes];
        for (sample, &label) in samples.iter().zip(labels.iter()) {
            if label < num_classes {
                class_samples[label].push(sample);
            }
        }

        // Compute class means
        let mut class_means: Vec<Vec<f32>> = Vec::new();
        for class_idx in 0..num_classes {
            let samples_for_class = &class_samples[class_idx];
            if samples_for_class.is_empty() {
                return Err(ClassifierError::EmptyClass(class_idx));
            }

            let mut mean = vec![0.0; num_features];
            for sample in samples_for_class {
                for (i, &val) in sample.iter().enumerate() {
                    mean[i] += val;
                }
            }
            for val in &mut mean {
                *val /= samples_for_class.len() as f32;
            }
            class_means.push(mean);
        }

        // Compute grand mean
        self.grand_mean = vec![0.0; num_features];
        for mean in &class_means {
            for (i, &val) in mean.iter().enumerate() {
                self.grand_mean[i] += val;
            }
        }
        for val in &mut self.grand_mean {
            *val /= num_classes as f32;
        }

        // Compute weights (simplified LDA)
        self.weights = class_means
            .iter()
            .map(|mean| {
                mean.iter()
                    .zip(self.grand_mean.iter())
                    .map(|(m, gm)| m - gm)
                    .collect()
            })
            .collect();

        self.is_trained = true;

        // Compute accuracy
        let mut correct = 0;
        for (sample, &label) in samples.iter().zip(labels.iter()) {
            let features = FeatureVector::new(0, sample.clone(), 4);
            let result = self.classify(&features);
            let predicted = self
                .classes
                .iter()
                .position(|g| *g == result.gesture)
                .unwrap_or(0);
            if predicted == label {
                correct += 1;
            }
        }

        let accuracy = correct as f32 / samples.len() as f32;

        Ok(TrainingResult {
            accuracy,
            training_duration_ms: start.elapsed().as_millis() as f32,
        })
    }

    /// Classify a feature vector
    pub fn classify(&self, features: &FeatureVector) -> ClassificationResult {
        let start = Instant::now();

        if !self.is_trained || self.weights.is_empty() {
            return ClassificationResult {
                gesture: Gesture::Rest,
                confidence: 0.0,
                alternatives: Vec::new(),
                latency_ms: start.elapsed().as_secs_f32() * 1000.0,
                timestamp: features.timestamp,
            };
        }

        // Compute discriminant scores
        let mut scores: Vec<f32> = Vec::new();
        for weights in &self.weights {
            let score: f32 = features
                .features
                .iter()
                .zip(self.grand_mean.iter())
                .zip(weights.iter())
                .map(|((&f, &gm), &w)| w * (f - gm))
                .sum();
            scores.push(score);
        }

        // Softmax
        let max_score = scores.iter().cloned().fold(f32::NEG_INFINITY, f32::max);
        let exp_scores: Vec<f32> = scores.iter().map(|&s| (s - max_score).exp()).collect();
        let sum_exp: f32 = exp_scores.iter().sum();
        let probabilities: Vec<f32> = exp_scores.iter().map(|&e| e / sum_exp).collect();

        // Find best class
        let (best_idx, &max_prob) = probabilities
            .iter()
            .enumerate()
            .max_by(|(_, a), (_, b)| a.partial_cmp(b).unwrap())
            .unwrap();

        // Build alternatives
        let mut alternatives: Vec<(Gesture, f32)> = self
            .classes
            .iter()
            .zip(probabilities.iter())
            .filter(|(_, &p)| p != max_prob)
            .map(|(&g, &p)| (g, p))
            .collect();
        alternatives.sort_by(|a, b| b.1.partial_cmp(&a.1).unwrap());
        alternatives.truncate(3);

        ClassificationResult {
            gesture: self.classes[best_idx],
            confidence: max_prob,
            alternatives,
            latency_ms: start.elapsed().as_secs_f32() * 1000.0,
            timestamp: features.timestamp,
        }
    }

    /// Check if classifier is trained
    pub fn is_trained(&self) -> bool {
        self.is_trained
    }

    /// Get supported gestures
    pub fn get_classes(&self) -> &[Gesture] {
        &self.classes
    }
}

/// Training result
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TrainingResult {
    pub accuracy: f32,
    pub training_duration_ms: f32,
}

/// Classifier errors
#[derive(Debug)]
pub enum ClassifierError {
    EmptyTrainingSet,
    EmptyClass(usize),
    NotTrained,
    InvalidFeatures,
}

impl std::fmt::Display for ClassifierError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            ClassifierError::EmptyTrainingSet => write!(f, "Empty training set"),
            ClassifierError::EmptyClass(idx) => write!(f, "Empty class: {}", idx),
            ClassifierError::NotTrained => write!(f, "Classifier not trained"),
            ClassifierError::InvalidFeatures => write!(f, "Invalid features"),
        }
    }
}

impl std::error::Error for ClassifierError {}

/// Majority voting buffer for temporal smoothing
pub struct MajorityVotingBuffer {
    buffer: Vec<ClassificationResult>,
    size: usize,
}

impl MajorityVotingBuffer {
    /// Create a new majority voting buffer
    pub fn new(size: usize) -> Self {
        Self {
            buffer: Vec::with_capacity(size),
            size,
        }
    }

    /// Add a result and get the majority vote
    pub fn add(&mut self, result: ClassificationResult) -> Option<ClassificationResult> {
        self.buffer.push(result);

        if self.buffer.len() > self.size {
            self.buffer.remove(0);
        }

        if self.buffer.len() < self.size {
            return None;
        }

        // Count votes
        let mut votes: HashMap<Gesture, (usize, f32)> = HashMap::new();
        for r in &self.buffer {
            let entry = votes.entry(r.gesture).or_insert((0, 0.0));
            entry.0 += 1;
            entry.1 += r.confidence;
        }

        // Find winner
        let (winner, (count, conf_sum)) = votes
            .iter()
            .max_by_key(|(_, (count, _))| *count)
            .unwrap();

        let avg_confidence = conf_sum / *count as f32;
        let avg_latency: f32 =
            self.buffer.iter().map(|r| r.latency_ms).sum::<f32>() / self.buffer.len() as f32;

        // Build alternatives
        let mut alternatives: Vec<(Gesture, f32)> = votes
            .iter()
            .filter(|(g, _)| *g != winner)
            .map(|(g, (c, s))| (*g, s / *c as f32))
            .collect();
        alternatives.sort_by(|a, b| b.1.partial_cmp(&a.1).unwrap());

        Some(ClassificationResult {
            gesture: *winner,
            confidence: avg_confidence,
            alternatives,
            latency_ms: avg_latency,
            timestamp: self.buffer.last().unwrap().timestamp,
        })
    }

    /// Clear the buffer
    pub fn clear(&mut self) {
        self.buffer.clear();
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_gesture_basic() {
        let gestures = Gesture::basic_gestures();
        assert_eq!(gestures.len(), 5);
        assert!(gestures.contains(&Gesture::Rest));
    }

    #[test]
    fn test_feature_vector() {
        let features = vec![0.1, 0.2, 0.3, 0.4, 0.5];
        let fv = FeatureVector::new(1234, features.clone(), 1);
        assert_eq!(fv.features, features);
        assert_eq!(fv.channel_count, 1);
    }

    #[test]
    fn test_lda_classifier() {
        let mut classifier = LDAClassifier::new(ClassifierConfig::default());

        // Create simple training data
        let samples = vec![
            vec![0.1, 0.1, 0.1, 0.1],
            vec![0.2, 0.1, 0.1, 0.1],
            vec![0.9, 0.9, 0.9, 0.9],
            vec![0.8, 0.9, 0.9, 0.9],
        ];
        let labels = vec![0, 0, 1, 1];
        let gestures = vec![Gesture::Rest, Gesture::HandClose];

        let result = classifier.train(&samples, &labels, gestures);
        assert!(result.is_ok());
        assert!(classifier.is_trained());
    }

    #[test]
    fn test_majority_voting() {
        let mut buffer = MajorityVotingBuffer::new(3);

        let r1 = ClassificationResult {
            gesture: Gesture::Rest,
            confidence: 0.8,
            alternatives: vec![],
            latency_ms: 1.0,
            timestamp: 1,
        };

        // First two shouldn't return anything
        assert!(buffer.add(r1.clone()).is_none());
        assert!(buffer.add(r1.clone()).is_none());

        // Third should return majority
        let result = buffer.add(r1);
        assert!(result.is_some());
        assert_eq!(result.unwrap().gesture, Gesture::Rest);
    }
}
